stumpy/cache.py

```diff
@@ -107,3 +109,40 @@ def _get_cache():
     site_pkg_dir = site.getsitepackages()[0]
     numba_cache_dir = site_pkg_dir + "/stumpy/__pycache__"
     return [f.name for f in pathlib.Path(numba_cache_dir).glob("*nb*") if f.is_file()]
+
+
+def _recompile(func=None, fastmath=None):
+    """
+    Recompile a jit/njit decorated function. If `func` is None, then it wiil
+    recompile all njit functions of STUMPY.
+
+    Parameters
+    ----------
+    func : a njit function, default None
+        The numba function to recompile. If None, then all njit functions
+        of STUMPY will be recompiled.
+
+    fastmath : bool or set, default None
+        The fastmath flags to use. If None, then the func's fastmath flags
+        will not be changed. This is only used when `func` is provided.
+
+    Returns
+    -------
+    None
+    """
+    warnings.warn(CACHE_WARNING)
+    if func is None:
+        njit_funcs = get_njit_funcs()
+        for module_name, func_name in njit_funcs:
+            module = importlib.import_module(f".{module_name}", package="stumpy")
+            func = getattr(module, func_name)
+            func.recompile()
+    else:
+        if not numba.extending.is_jitted(func):
```

Comment:
I am not checking if `func` is one of the STUMPY's njit functions. I am allowing this function to recompile any njit function.

Reply:
This is probably not a good idea as allowing the recompilation of ANY NJIT function (even outside of STUMPY) will have unintended consequences. It is trivial to add `if func in njit_funcs:`.

I also don't like how this is a nested `if/else`. I think it can be simplified. For example, you might be able to do something like:

```
njit_funcs = get_njit_funcs()
recompile_funcs = []

# Collect list of functions for recompiling
if func in njit_funcs:
    recompile_funcs.append(func)
# You may add a `elif func not in njit_funcs` if you want
else:
    for module_name, func_name in njit_funcs:
        module = importlib.import_module(f".{module_name}", package="stumpy")
        func = getattr(module, func_name)
        recompile_funcs.append(func)

# Recompile functions
for func in recompile_funcs():
    if fastmath is not None:
            func.targetoptions["fastmath"] = fastmath
    func.recompile()
```

I think this logic is much, much simpler to follow. Additionally, it limits functions to those in STUMPY. I wonder if we should/could refactor the other functions to look/behave similarly

Reply:
Thanks for the feedback! I intentionally wanted to provide support for recompiling ANY njit function as I thought that would help users :) I remember you once said it is sometimes better to limit the scope of a function! This is one of those cases! 

---

> \# You may add a `elif func not in njit_funcs` if you want

```
recompile_funcs = []
if func is None:
     # append all STUMPY's njit functions to recompile_funcs
elif func is njit_funcs:
    # append that func to recompile_funcs
else:
    # raise error that the provided function is not of STUMPY's.

# Recompile funcs in recompile_funcs
```

---

@seanlaw 
> I wonder if we should/could refactor the other functions to look/behave similarly

The other functions in `stumpy/cache.py`? to add a support that enables users to apply them to only one STUMPY's function or all of them?

Reply:
> I intentionally wanted to provide support for recompiling ANY njit function as I thought that would help users :) I remember you once said it is sometimes better to limit the scope of a function! This is one of those cases!

Yes! In fact, the only reason why we are forced to do this is because this functionality hasn't been exposed directly in `numba` (or there isn't obvious API). However, STUMPY should only be responsible for STUMPY itself and not for `numba`, `numpy`, or `scipy`! If you were really motivated, you should add a PR to `numba` to offer the functionality and, once implemented, you can come back to STUMPY and replace the hacked functionality with the `numba`-approved function.

> The other functions in stumpy/cache.py? to add a support that enables users to apply them to only one STUMPY's function or all of them?

Yes! Basically, this would be adding a `func=None` parameter to both `cache._enable()` and `cache._clear()`. Currently, both of these functions offer all-or-nothing cache enabling/clearing.

Reply:
Created #1056 to track it later.